pyproject.toml

```diff
@@ -0,0 +1,66 @@
+[build-system]
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "uncertainties"
+version = "3.1.7"
+authors = [
+    {name = "Eric O. LEBIGOT (EOL)", email = "eric.lebigot@normalesup.org"},
+]
+description = """\
+    Transparent calculations with uncertainties on the \
+    quantities involved (aka error propagation); \
+    fast calculation of derivatives').\
+    """
+readme = "README.rst"
+requires-python = ">=2.7"
+keywords = [
+    "error propagation", "uncertainties", "uncertainty calculations",
+    "standard deviation", "derivatives", "partial derivatives",
+    "differentiation"
+]
+license = {text = "Revised BSD License"}
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Education",
+    "Intended Audience :: Other Audience",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: BSD License",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.1",
```

Comment:
should the python <3.8 be included here?

Reply:
the development plans discussions lists 
> 1. Support Python 3.8 to 3.12

So you could remove the future dependency and early python versions here, or that can be done in another PR.

Reply:
Yes, it seemed like there was agreement to move to 3.8+. I started to do that here but then thought I would do it in a separate PR since it is a consequential change and it would be nice to have its own PR for it for someone to comment on. I think we can just make a quick follow up to drop the lines below 3.8.

There is also the question of `requires-python`. pip obeys that line, so we may not want to set it to `>=3.8` if we want to allow Python 3.6 until we intentionally break compatibility, even if we are not actively testing it. I left it at 2.7 here, but I think this change breaks the ability to install the package in Python 2 already since I don't think the last versions of pip and setuptools for Python 2 can install a project using pyproject.toml.

Maybe I should make a PR to drop Python 2 and early Python 3 to merge before this PR so there is a clear progression. I did this part first because I thought it was easier (not needing decide what line to draw with 3.6/3.8).

Reply:
I just quickly researched how `requires-python` works here. It looks like if we set `requires-python` to `>=3.8` then if someone did `pip install uncertainties` while having Python 3.7 then pip/pypi would NOT let them install the newest version of uncertainties, but it would walk backwards finding older versions until one did not have 3.7 excluded by `requires-python`. 

I'm on the fence about if we want to set `requires-python` or not. I guess I agree with you we shouldn't set it until we know we've broken compatibility. But I guess I would suggest that we try to set it **exactly** when we do know we break compatibility. That is, if we know the code is broken for Python 3.7 we should set `requires-python = ">=3.8"` at that time. That way users hit the error as soon as possible (i.e. at install time). I guess what would happen for that user is pip/pypi would resolve an old enough version of `uncertainties` that is compatible with their python version. This combined with messaging in the docs about which versions of `uncertainties` support which versions of Python seems like enough notification to users about versioning considerations.